Spigot-Server-Patches/0334-Adding-alternate-and-more-even-spawn-mechanic.patch

```diff
@@ -0,0 +1,547 @@
++        if(AMBIENT.getInnerClass().isAssignableFrom(clazz)) return AMBIENT;
++        if(WATER_CREATURE.getInnerClass().isAssignableFrom(clazz)) return WATER_CREATURE;
++
++        return null;
++    }
++    // Paper end
++
++    public Class<? extends IAnimal> getInnerClass() { return a(); } // Paper - OBFHELPER
+     public Class<? extends IAnimal> a() {
+         return this.e;
+     }
+ 
++    public int getSpawnLimit() { return b(); } // Paper - OBFHELPER
+     public int b() {
+         return this.f;
+     }
+ 
++    public boolean isPassive() { return c(); } // Paper - OBFHELPER
+     public boolean c() {
+         return this.g;
+     }
+ 
++    public boolean isRare() { return d(); } // Paper - OBFHELPER
+     public boolean d() {
+         return this.h;
+     }
+diff --git a/src/main/java/net/minecraft/server/PlayerChunk.java b/src/main/java/net/minecraft/server/PlayerChunk.java
+index fcd9f5491..6f04d4759 100644
+--- a/src/main/java/net/minecraft/server/PlayerChunk.java
++++ b/src/main/java/net/minecraft/server/PlayerChunk.java
+@@ -249,6 +249,7 @@ public class PlayerChunk {
+         return false;
+     }
+ 
++    public boolean isDone() { return e(); } // Paper - OBFHELPER
+     public boolean e() {
+         return this.done;
+     }
+diff --git a/src/main/java/net/minecraft/server/SpawnerCreature.java b/src/main/java/net/minecraft/server/SpawnerCreature.java
+index f525fd1b4..23a0fc6a3 100644
+--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
++++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
+@@ -1,10 +1,12 @@
+ package net.minecraft.server;
+ 
+-import com.google.common.collect.Sets;
+ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Random;
+ import java.util.Set;
++import java.util.HashSet; // Paper
++import java.util.Map;
++import java.util.EnumMap; // Paper
+ import javax.annotation.Nullable;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
+@@ -21,10 +23,288 @@ public final class SpawnerCreature {
+     private static final Logger a = LogManager.getLogger();
+     private static final int b = (int) Math.pow(17.0D, 2.0D);
+     private final LongHashSet c = new LongHashSet(); // CraftBukkit
++    public static int seed = "paper".hashCode(); // Paper
+ 
+     public SpawnerCreature() {}
+ 
++    public static class ChunkCoordOrdinalInt3Tuple extends ChunkCoordIntPair {
++        public int ordinal;
++        public int cachedHashCode = -1;
++
++        public static int mix(int hash, int data) {
++            int k = data;
++
++            k *= 0xcc9e2d51;
++            k = (k << 15) | (k >>> -15);
++            k *= 0x1b873593;
++
++            int h = hash ^ k;
++
++            h = (h << 13) | (h >>> -13);
++
++            return h * 5 + 0xe6546b64;
++        }
++
++        public static int createHashCode(int ordinal, int x, int z) {
++            //MurmurHash3_x86_32 hashing
```

Comment:
Instead of copying the algorithm, you could use the Guava implementation instead: https://github.com/google/guava/blob/f264e9671e33744b4b7733cf3e4c8eb95be51999/guava/src/com/google/common/hash/Hashing.java#L47-L134